core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -87,12 +91,13 @@ abstract class AbstractFetcherThread(name: String,
     fetcherLagStats.unregister()
   }
 
-  override def doWork() {
+  def states() = partitionStates.partitionStates.asScala.map { state => state.topicPartition -> state.value}
 
+  override def doWork() {
     val fetchRequest = inLock(partitionMapLock) {
-      val fetchRequest = buildFetchRequest(partitionStates.partitionStates.asScala.map { state =>
-        state.topicPartition -> state.value
-      })
+      initialisePartitions(states)
```

Comment:
I've pulled the logic up into AbstractFetcherThread. What I've done is somewhat similar to your snippet, but it includes a lock around the truncation step. 

 ```
 override def doWork() {
    maybeTruncate()
...
...
  def maybeTruncate(): Unit = {
    val epochRequests = inLock(partitionMapLock) {buildLeaderEpochRequest(states)}
    if (!epochRequests.isEmpty) {
      val fetchedEpochs = fetchEpochsFromLeader(epochRequests)

      //Ensure we hold a lock during truncation.
      inLock(partitionMapLock) {
        //Check no leadership changes happened whilst we were unlocked, fetching epochs
        val leaderEpochs = fetchedEpochs.filter { case (tp, _) => partitionStates.contains(tp) }
        val truncationPoints = maybeTruncate(leaderEpochs)
        markInitialisationComplete(truncationPoints)
      }
    }
  }
```